halo2_gadgets/src/utilities.rs

```diff
@@ -59,6 +78,62 @@ pub trait UtilitiesInstructions<F: FieldExt> {
     }
 }
 
+/// A type representing a range-constrained field element.
+#[derive(Clone, Copy, Debug)]
+pub struct RangeConstrained<F: Field, T: FieldValue<F>> {
+    inner: T,
+    num_bits: usize,
+    _phantom: PhantomData<F>,
+}
+
+impl<F: Field, T: FieldValue<F>> RangeConstrained<F, T> {
+    /// Returns the range-constrained inner type.
+    pub fn inner(&self) -> &T {
+        &self.inner
+    }
+
+    /// Returns the number of bits to which this cell is constrained.
+    pub fn num_bits(&self) -> usize {
+        self.num_bits
+    }
+}
+
+impl<F: PrimeFieldBits> RangeConstrained<F, Option<F>> {
+    /// Constructs a `RangeConstrained<Option<F>>` as a subset of the given value.
+    pub fn subset_of(value: Option<&F>, bitrange: Range<usize>) -> Self {
+        let num_bits = bitrange.len();
+        Self {
+            inner: value.map(|value| bitrange_subset(value, bitrange)),
+            num_bits,
+            _phantom: PhantomData::default(),
+        }
+    }
+}
+
+impl<F: Field> RangeConstrained<F, AssignedCell<F, F>> {
+    /// Constructs a `RangeConstrained<AssignedCell<F, F>>` without verifying that the
+    /// cell is correctly range constrained.
+    ///
+    /// This API only exists to ease with integrating this type into existing circuits,
+    /// and will likely be removed in future.
```

Comment:
Unsure if it's appropriate here to set `#[doc(hidden)]` if it's intended to be removed.

Reply:
My plan is to finish going through the circuit and finding all the ways we want to construct this type, and then see how we can better integrate it into the existing APIs.

Reply:
I'm going to come back to this in a second PR going over the `halo2_gadgets` circuit impls.